src/compiler/checker.ts

```diff
@@ -9428,6 +9504,18 @@ namespace ts {
         }
 
         function checkApplicableSignature(node: CallLikeExpression, args: Expression[], signature: Signature, relation: Map<RelationComparisonResult>, excludeArgument: boolean[], reportErrors: boolean) {
+            const headMessage = Diagnostics.Argument_of_type_0_is_not_assignable_to_parameter_of_type_1;
+            if (signature.thisType && signature.thisType !== voidType && node.kind !== SyntaxKind.NewExpression) {
+                // If the source is not of the form `x.f`, then sourceType = voidType
+                // If the target is voidType, then the check is skipped -- anything is compatible.
+                // If the the expression is a new expression, then the check is skipped.
+                const calleeNode = node.kind === SyntaxKind.CallExpression && (<PropertyAccessExpression>(<CallExpression>node).expression).expression;
+                const calleeType: Type = calleeNode ? checkExpressionWithContextualType(calleeNode, signature.thisType, undefined) : voidType;

```

Comment:
The expression in the callee node has already been type checked by the time we get here (in a `x.f(...)` type of call we need to know the type of `x` before we can even figure out what `f` refers to), so it doesn't make sense to say that the callee node is contextually typed by the this type.